### ISSUE TITLE
Survey Workflow: design

### ISSUE THREAD

**Issue by JWoodland-Scott** [finos/waltz#5757]
Users would like the ability to use a workflow system to drive the issuance of surveys with a view to updating one or more assessments in waltz. 

For example, an architectural review process:
This workflow is used to track architectural governance compliance of change initiatives.

2 assessments will be used:
- Outcome assessment
- Workflow status assessment

An external batch job determines which change initiatives are in scope and sets the outcome assessment to 'In Scope'.
A workflow job will wait for assessments in this state and issue a "scoping" survey and update the workflow status assessment to be 'Pending Completion'. etc.... 
Once the survey is approved the initial assessment is updated to reflect the end state.

**Comment by JWoodland-Scott** [comment]
###workflow_definition

Name                        |       Type            | Description
---                         | ---                   | ---
id                          | Long                  | Sequential identifier
name                        | String                | name
description                 | String                | description
external_id                 | String                | external id
created_at                  | Timestamp             | workflow first created
created_by                  | String                | user to create workflow
owning_role                 | String                | which users can modify this workflow
status                      | String                | ACTIVE / DRAFT / REMOVED


###workflow_state
Describes the different states of a workflow, the state_kind indicates the stage of this state in the workflow

Name                 |       Type               | Description
---                  | ---                      | ---
id                   |       Long               | identifier
workflow_defn_id     |       Long               | FK to workflow_definition
name                 |       String             | name
description          |       String             | description
external_id          |       String             | Description
state_kind           |       String             | NOT_STARTED/ IN_PROGRESS /COMPLETED


###workflow_instance

Name                    |       Type            | Description
---                     | ---                   | ---
id                      | Long                  | Sequential identifier
workflow_definition_id  | Long                  | FK to workflow definition
state_id                | Long                  | state of this instance
parent_entity_id        | Long                  | Entity this workflow is acting on
parent_entity_kind      | String                | Entity this workflow is acting on
external_id             | String                | external id 
created_at              | Timestamp             | workflow first created
created_by              | String                | user to create workflow
last_updated_at         | Timestamp             | when was this instance last updated
last_updated_by         | String                | user to last update this instance

**Comment by JWoodland-Scott** [comment]
There are the following tables in Waltz which could be modified to accommodate these changes: 

- entity_workflow_definition
- entity_workflow_state
- entity_workflow_transition

Most likely by creating a new table to track the state, and renaming entity_workflow_state to entity_workflow_instance

**Comment by JWoodland-Scott** [comment]
Proposed tables:

## workflow_definition

Attribute | Type | Description
-- | -- | --
id | Long | Sequential identifier
name | String | name
description | String | description
external_id | String | external id
created_at | Timestamp | workflow first created
created_by | String | user to create workflow
owning_role | String | which users can modify this workflow
status | String | ACTIVE / DRAFT / REMOVED

## workflow_state

Attribute | Type | Description
-- | -- | --
id | Long | identifier
workflow_defn_id | Long | FK to workflow_definition
name | String | name
description | String | description
external_id | String | Description
state_kind | String | NOT_STARTED/ IN_PROGRESS /COMPLETED

## workflow_instance

Attribute | Type | Description
-- | -- | --
id | Long | Sequential identifier
workflow_definition_id | Long | FK to workflow definition
state_id | Long | state of this instance
parent_entity_id | Long | Entity this workflow is acting on
parent_entity_kind | String | Entity this workflow is acting on
external_id | String | external id
created_at | Timestamp | workflow first created
created_by | String | user to create workflow
last_updated_at | Timestamp | when was this instance last updated
last_updated_by | String | user to last update this instance

**Comment by JWoodland-Scott** [comment]
- [ ] setup ddl
  - [ ] migrate current entity_workflow_definition' to 'workflow_definition' (adding external id and owner, status etc...)
  - [ ] introduce 'workflow_state' table and populate with the distinct list of state values form the existing 'entity_workflow_state' table
  - [ ] migrate 'entity_workflow_state' to 'workflow_instance' (state -> state_id etc)
  - [ ] migrate 'entity_workflow_transition' to 'workflow_transition' (drop entity columns, use instance_id and state_id).
  - [ ] update existing internal jobs to use the new tables
- [ ] Update / recreate model for new workflow tables
- [ ] Workflow job
  - [ ] Externalise rules 
  - [ ] Rule evaluator
  - [ ] Define predicates for rules
  - [ ] Define 'side-effects' for rules e.g. issue surveys or update assessment
- [ ] Workflow admin screens

**Comment by JWoodland-Scott** [comment]
-- Records Mmt

Initialisation
	switch 
		case:
			- Pred: App in development && has no survey issued && is not euda
			- State: INITIAL_SURVEY_ISSUED/IN_PROGRESS
			- Side-effect: issueSurvey("INIT")
		default:
			- State: NOT_REQUIRED/PARKED

Transitions

    HasRecords
		- InitialState: INITIAL_SURVEY_ISSUED
		- Pred: hasApprovedSurvey("INIT") && surveyAnswer("INIT", "Q1") == "HAS_RECORDS"
		- State: DETAIL_SUREVY_ISSUED/IN_PROGRESS
		- Side-effect: issueSurveyToAllEntitiesListedInSurvey("DETAIL", "INITIAL", "Q3") and updateAssessment("HAS_RECORDS", "YES")
	
	DoesNotHaveRecords
		- InitialState: INITIAL_SURVEY_ISSUED
		- Pred: hasApprovedSurvey("INIT") && surveyAnswer("INIT", "Q1") == "DOES_NOT_HAVE_RECORDS"
		- State: NO_RECORDS/COMPLETED
		- Side-effect: updateAssessment("HAS_RECORDS", "NO")
		
	DetailSurveyComplete
		- InitialState: DETAIL_SUREVY_ISSUED
		- Pred: hasApprovedSurvey("DETAIL")
		- State: DETAIL_SURVEY_COMPLETED/COMPLETED
		- Side-effect: updateRatingsFromSurveyQuestion("RECORDS", "12", "YES")
		
		
		
-- Arch Gov

Initialisation
	switch 
		case:
			- Pred: CI is not retired and hasAssessment("ARCH_SCOPE") == 'IN_SCOPE' (from clarity checks)
			- State: DISCOVERY_SURVEY_ISSUED/IN_PROGRESS
			- Side-effect: issueSurvey("DISCOVERY")
		default:
			- State: NOT_REQUIRED/PARKED

Transitions

    NoArchChange
		- InitialState: DISCOVERY_SURVEY_ISSUED
		- Pred: hasApprovedSurvey("DISCOVERY") && surveyAnswer("DISCOVERY", "Q1") == "No - No Architectural Change"
		- State: DISCOVER_SURVEY_NO_ARCH_CHANGE/COMPLETED
		- Side-effect: updateAssessment("ARCH_SCOPE", "DESCOPED")
		
    ArchChange
		- InitialState: DISCOVERY_SURVEY_ISSUED
		- Pred: hasApprovedSurvey("DISCOVERY") && surveyAnswer("DISCOVERY", "Q1") == "Yes"
		- State: ARCH_REVIEW_SURVEY_ISSUED/IN_PROGRESS
		- Side-effect: issueSurvey("ARCH_REVIEW")
		
	DiscoveryOverdue
		- InitialState: DISCOVERY_SURVEY_ISSUED
		- Pred: !hasApprovedSurvey("DISCOVERY") && monthsSinceLastUpdate > 6
		- State: DISCOVERY_SURVEY_OVERDUE/IN_PROGRESS
		- Side-effect: issueReminder()
		
	 ArchChangeFromOverdue
		- InitialState: DISCOVERY_SURVEY_OVERDUE
		- Pred: hasApprovedSurvey("DISCOVERY")
		- State: DISCOVERY_SURVEY_ISSUED/IN_PROGRESS		

	ArchReviewCompleted
		- InitialState: ARCH_REVIEW_SURVEY_ISSUED
		- Pred: hasApprovedSurvey("ARCH_REVIEW")
		- State: ARCH_REVIEW_COMPLETED/COMPLETED
		- Side-effect: updateAssessment("ARCH_SCOPE", getValueFromSurveyQuestion("ARCH_REVIEW", "Q4"))
